SimPEG/DataMisfit.py

```diff
@@ -1,65 +1,45 @@
 from __future__ import print_function
 import numpy as np
-from SimPEG import Utils
-from SimPEG import Survey
+from . import Utils
+from . import Survey
+from . import ObjectiveFunction
+from .Tests import checkDerivative
 
 
-class BaseDataMisfit(object):
-    """BaseDataMisfit
+class BaseDataMisfit(ObjectiveFunction.L2ObjectiveFunction):
+    """
+    BaseDataMisfit
 
         .. note::
 
-            You should inherit from this class to create your own data misfit term.
+            You should inherit from this class to create your own data misfit
+            term.
     """
 
     debug   = False  #: Print debugging information
-    counter = None   #: Set this to a SimPEG.Utils.Counter() if you want to count things
+    counter = None  #: Set this to a SimPEG.Utils.Counter() if you want to count things
 
     def __init__(self, survey, **kwargs):
         assert survey.ispaired, 'The survey must be paired to a problem.'
         if isinstance(survey, Survey.BaseSurvey):
             self.survey = survey
             self.prob   = survey.prob
-        Utils.setKwargs(self,**kwargs)
-
-    @Utils.timeIt
-    def eval(self, m, f=None):
-        """eval(m, f=None)
-
-            :param numpy.array m: geophysical model
-            :param Fields f: fields
-            :rtype: float
-            :return: data misfit
-
-        """
-        raise NotImplementedError('This method should be overwritten.')
-
-    @Utils.timeIt
-    def evalDeriv(self, m, f=None):
-        """evalDeriv(m, f=None)
-
-            :param numpy.array m: geophysical model
-            :param Fields f: fields
-            :rtype: numpy.array
-            :return: data misfit derivative
+        super(BaseDataMisfit, self).__init__(**kwargs)
 
-        """
-        raise NotImplementedError('This method should be overwritten.')
-
-
-    @Utils.timeIt
-    def eval2Deriv(self, m, v, f=None):
-        """eval2Deriv(m, v, f=None)
-
-            :param numpy.array m: geophysical model
-            :param numpy.array v: vector to multiply
-            :param Fields f: fields
-            :rtype: numpy.array
-            :return: data misfit derivative
-
-        """
-        raise NotImplementedError('This method should be overwritten.')
+    @property
+    def nP(self):
+        if self._mapping is not None:
+            return self.mapping.nP
+        elif self.prob.model is not None:
+            return len(self.prob.model)
+        else:
+            return '*'
 
+    @property
```

Comment:
it might be better to do a soft depreciation on this - allow there to be a setter that throws a warning and sets W instead of Wd. thoughts @rowanc1 ?